core/src/main/java/com/google/instrumentation/common/TimestampFactory.java

```diff
@@ -13,16 +13,52 @@
 
 package com.google.instrumentation.common;
 
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.Immutable;
+
 /**
- * An interface for a factory that generates timestamps.
- *
- * @see Timestamp
+ * Factory for {@link Timestamp}. See {@link #now} for how to use this class.
  */
-public interface TimestampFactory {
+public final class TimestampFactory {
+  private static final Logger logger = Logger.getLogger(TimestampFactory.class.getName());
+  private static final Handler HANDLER =
+      loadTimestamFactoryHandler(Provider.getCorrectClassLoader(Handler.class));
+
+  private TimestampFactory() {}
+
   /**
-   * Returns a new timestamp that represents the instant this method was called.
+   * Obtains the current instant from the system clock.
    *
-   * @return the new timestamp.
+   * @return the current instant using the system clock, not null.
    */
-  Timestamp now();
+  public static Timestamp now() {
+    return HANDLER.timeNow();
+  }
+
+  // This class cannot be final because in case of java8 runtime we can have a better granularity
+  // timestamp.
+  @Immutable
+  static class Handler {
```

Comment:
I think that it would be slightly cleaner to split this class into an interface and an immutable final class.  Then the Java 8 class could implement the interface.